sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsReadableSpan.java

```diff
@@ -159,24 +160,29 @@ static RecordEventsReadableSpan startSpan(
   @Override
   public SpanData toSpanData() {
     SpanContext spanContext = getSpanContext();
-    return SpanData.newBuilder()
-        .setName(getName())
-        .setInstrumentationLibraryInfo(instrumentationLibraryInfo)
-        .setTraceId(spanContext.getTraceId())
-        .setSpanId(spanContext.getSpanId())
-        .setTraceFlags(spanContext.getTraceFlags())
-        .setTracestate(spanContext.getTracestate())
-        .setAttributes(getAttributes())
-        .setStartEpochNanos(startEpochNanos)
-        .setEndEpochNanos(getEndEpochNanos())
-        .setKind(kind)
-        .setLinks(getLinks())
-        .setParentSpanId(parentSpanId)
-        .setHasRemoteParent(hasRemoteParent)
-        .setResource(resource)
-        .setStatus(getStatus())
-        .setTimedEvents(adaptTimedEvents())
-        .build();
+    // Set immutable info outside the lock.
+    SpanData.Builder builder =
+        SpanData.newBuilder()
+            .setInstrumentationLibraryInfo(instrumentationLibraryInfo)
+            .setTraceId(spanContext.getTraceId())
+            .setSpanId(spanContext.getSpanId())
+            .setTraceFlags(spanContext.getTraceFlags())
+            .setTracestate(spanContext.getTracestate())
+            .setStartEpochNanos(startEpochNanos)
+            .setKind(kind)
+            .setLinks(getLinks())
+            .setParentSpanId(parentSpanId)
+            .setHasRemoteParent(hasRemoteParent)
+            .setResource(resource);
+    synchronized (lock) {
+      builder
+          .setName(getName())
+          .setAttributes(hasBeenEnded ? attributes : new HashMap<>(attributes))
```

Comment:
I'm not sure if HashMap is correct here, since we used a LinkedHashMap in the span. But from my understanding this is only required to remove the right attribute if too many are added.